clippy_lints/src/loops/manual_find.rs

```diff
@@ -0,0 +1,145 @@
+                            ExprKind::MethodCall(_, [caller, ..], _) if is_local_used(cx, caller, binding_id)
+                        ) && is_local_used(cx, side, binding_id)
+                        {
+                            needs_copy = true;
+                        }
+                    }
+                },
+                ExprKind::Call(..) => needs_copy = true,
+                ExprKind::MethodCall(_, [caller, ..], _) => {
+                    if !is_local_used(cx, caller, binding_id) {
+                        needs_copy = true;
+                    }
+                },
+                _ => {},
+            }
+            let ty = cx.typeck_results().expr_ty(inner_ret);
+            if !needs_copy || cx.tcx.lang_items().copy_trait().map_or(false, |id| implements_trait(cx, ty, id, &[])) {
+                snippet.push_str(
+                    &format!(
+                        ".find(|{}{}| {})",
+                        "&".repeat(usize::from(needs_copy) + usize::from(is_ref_to_binding)),
+                        snippet_with_applicability(cx, inner_ret.span, "..", &mut applicability),
+                        snippet_with_applicability(cx, cond.span, "..", &mut applicability),
+                    )[..],
+                );
+                if is_ref_to_binding {
+                    snippet.push_str(".copied()");
+                }
+            } else {
+                applicability = Applicability::HasPlaceholders;
+                snippet.push_str(
+                    &format!(
+                        ".find(|{}| {{ ... }})",
+                        snippet_with_applicability(cx, inner_ret.span, "..", &mut applicability),
+                    )[..],
+                );
+            }
+            span_lint_and_sugg(
+                cx,
+                MANUAL_FIND,
+                span.to(last_ret.span),
+                "manual implementation of Iterator::find",
+                "replace with an iterator",
+                snippet,
+                applicability,
+            );
+        }
+    }
+}
+
+fn get_binding(pat: &Pat<'_>) -> Option<HirId> {
+    let mut hir_id = None;
+    let mut count = 0;
+    pat.each_binding(|annotation, id, _, _| {
+        count += 1;
+        if count > 1 {
+            hir_id = None;
+            return;
+        }
+        if let BindingAnnotation::Unannotated = annotation {
+            hir_id = Some(id);
+        }
+    });
+    hir_id
+}
+
+// Returns last_return if function fits format for lint
+fn last_ret<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> Option<&'tcx Expr<'tcx>> {
+    let mut parent_iter = cx.tcx.hir().parent_iter(expr.hir_id);
+    let mut last_return = None;
+    if_chain! {
+        if let Some((node_hir, Node::Stmt(..))) = parent_iter.next();
+        if let Some((_, Node::Block(block))) = parent_iter.next();
+        if let [.., last_stmt] = block.stmts;
+        if last_stmt.hir_id == node_hir;
+        if let Some(last_ret) = block.expr;
```

Comment:
This won't cover `return None` as the last block expression/statement. I suggest to move the `strip_return` function in `matches::needless_match` to `clippy_utils` and reuse it here.

Please also add a test for this then.